schism/page_patedit.c

```diff
@@ -2958,28 +2960,50 @@ static int patedit_record_note(song_note_t *cur_note, int channel, UNUSED int ro
 static int pattern_editor_insert_midi(struct key_event *k)
 {
 	song_note_t *pattern, *cur_note = NULL;
-	int n, v = 0, c = 0, pd, speed, tick;
+	int n, v = 0, pd, speed, tick, offset = 0;
+	int r = current_row, c = current_channel, p = current_pattern;
+	int first_note = 0;
+	int quantize_next_row = 0;
+	int ins = KEYJAZZ_NOINST, smp = KEYJAZZ_NOINST;
+
+	if (song_is_instrument_mode()) {
+		ins = instrument_get_current();
+	} else {
+		smp = sample_get_current();
+	}
 
 	status.flags |= SONG_NEEDS_SAVE;
-	song_get_pattern(current_pattern, &pattern);
 
-	if (midi_start_record && !(song_get_mode() & (MODE_PLAYING|MODE_PATTERN_LOOP))) {
+	if (midi_start_record && !SONG_PLAYING) {
 		switch (midi_start_record) {
 		case 1: /* pattern loop */
-			song_loop_pattern(current_pattern, current_row);
+			song_loop_pattern(p, r);
 			midi_playback_tracing = playback_tracing;
 			playback_tracing = 1;
 			break;
 		case 2: /* song play */
-			song_start_at_pattern(current_pattern, current_row);
+			song_start_at_pattern(p, r);
 			midi_playback_tracing = playback_tracing;
 			playback_tracing = 1;
 			break;
 		};
+
+		first_note = 1;
 	}
 
+	/* correct late notes to the next row */
+	/* tick + 1 because processing the keydown itself takes another tick */
+	if (midi_flags & MIDI_TICK_QUANTIZE && SONG_PLAYING && !first_note
```

Comment:
This the only place `first_note` is used, so making this conditional an `else if` of the previous one would be simpler and easier to understand.

Reply:
guess you're right. fixing.

Reply:
there is still a bug here. what if `midi_start_record` is zero and the song is playing? keyjazz will be quantized to next row based on tick, even if the trigger mode (ctrl-z) is none. maybe add `&& midi_start_record`